pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -492,7 +492,14 @@ def _binary_op_method_timedeltalike(op, name):
             if other.dtype.kind not in ['m', 'M']:
                 # raise rathering than letting numpy return wrong answer
                 return NotImplemented
-            return op(self.to_timedelta64(), other)
+            result = op(self.to_timedelta64(), other)
+            if other.ndim == 0:
```

Comment:
Wouldn't be cleaner to do `elif hasattr(other, 'dtype') and other.ndim > 0`, and then let the scalar numpy ones take the route of Timedelta below ? (the `other = Timedelta(other)` some lines below will work for a np.timedelta64

Reply:
I think that would be about a wash because we also need to handle the case where other is a 0-dim datetime64 array.

Reply:
Isn't that already catched by the `is_datetime64_object(other)` above? (not sure what the method exactly does, but from the name I would expect that)

Reply:
And from this it seems it is indeed already working correctly:

```
In [163]: pd.Timedelta(1) + np.datetime64(1000, 'ns')
Out[163]: Timestamp('1970-01-01 00:00:00.000001001')
```

Reply:
Yah there's some ambiguity there.  The 0-dim case to watch out for is (under master, fixed in the PR):

```
>>> pd.Timedelta(1) + np.array('2016-01-02 03:04:05', dtype='datetime64[ns]')
numpy.datetime64('2016-01-02T03:04:05.000000001')
```

Reply:
Ah, but not sure that *should* be fixed? Since now it would depend on the dimension of the array what the return type is. 
I certainly agree to fix this case:

```
In [169]: pd.Timedelta(1) + np.timedelta64(1)
Out[169]: numpy.timedelta64(2,'ns')
```

but not sure we should add that much extra lines to just deal with numpy scalars vs numpy 0-dim arrays

Reply:
> Ah, but not sure that should be fixed? 

I could see a reasonable argument that the zero-dim array op should return a zero-dim array (i.e. not a Timedelta like the PR changes it to), but given that it returns a scalar, I think the consistent thing to do is always return a Timestamp/Timedelta.

> not sure we should add that much extra lines to just deal with numpy scalars vs numpy 0-dim arrays

As long as we a) aren't making spaghetti and b) are testing these new corner cases, I don't see much downside to fixing these.  If there were some kind of LOC budget I'd agree with you that this would be a pretty low priority.  (and if it will get this merged I'll revert this part, since this is blocking fixes to Index bugs)

Reply:
> . (and if it will get this merged I'll revert this part, since this is blocking fixes to Index bugs)

Can you explain why this part would be reverted again, but is needed to be first merged?

Reply:
Poor use of pronouns on my part.  "This part" refers to the zero-dimensional arrays, and is _not_ the bug that originally motivated this PR, but was found in the process of writing tests for this PR.  "this is blocking" referred to the divmod/mod and 1-d array ops, which I need to have here before I can fix e.g. `pd.Index([0, 1, 2]) * pd.Timedelta(days=1)`

Reply:
I don't see any test you added that would catch this 0-dim array case. Would it be ok to just leave that out of this PR, and only fix the numpy timedelta64 scalar case?

We can discuss how many lines of code are added and whether that is worth it, but checking the result type and the potentially converting the result, *does* add to the code complexity. I would rather add a `is_timedelta64_object(other)` check to explicitly use the normal timedelta path for those.

Reply:
I'm amenable to this suggestion, will update in a bit.

FYI `is_timedelta64_object(obj)` is equivalent to isinstance(obj, np.timedelta64).  It will _not_ catch arrays with timedelta64-dtype.